components/core/src/clp/Grep.cpp

```diff
@@ -1063,4 +1044,468 @@ size_t Grep::search(Query const& query, size_t limit, Archive& archive, File& co
+        // simultaneously match multiple variables and static text, and we need a different
+        // approach to compare against the archive.
+        bool contains_wildcard = false;
+        set<uint32_t> variable_types;
+        if (has_preceding_delimiter && has_proceeding_delimiter) {
+            // If the substring is preceded or proceeded by a greedy wildcard then it's
+            // possible the substring could be extended to match a var, so the wildcards are
+            // added to the substring. If we don't consider this case we could miss
+            // combinations. Take for example "a*b", "a*" and "*b" can both match a has#
+            // style variable ("\w*\d+\w*"). If we decompose the string into either
+            // substrings "a*" + "b" or "a" + "*b", neither would capture the possibility of
+            // a logtype with the form "<has#>*<has#>", which is a valid possibility during
+            // compression. Instead we desire to decompose the string into "a*" + "*" +
+            // "*b". Note, non-greedy wildcards do not need to be considered, for example
+            // "a?b" can never match "<has#>?<has#>" or "<has#><has#>".
+            uint32_t substr_start = begin_idx;
+            uint32_t substr_end = end_idx;
+            bool prev_char_is_star = begin_idx > 0 && is_greedy_wildcard[begin_idx - 1];
+            bool next_char_is_greedy_wildcard
+                    = end_idx < processed_search_string.length() && is_greedy_wildcard[end_idx];
+            if (prev_char_is_star) {
+                substr_start--;
+            }
+            if (next_char_is_greedy_wildcard) {
+                substr_end++;
+            }
+            std::tie(variable_types, contains_wildcard) = get_substring_variable_types(
+                    string_view(processed_search_string)
+                            .substr(substr_start, substr_end - substr_start),
+                    substr_start,
+                    is_greedy_wildcard,
+                    is_non_greedy_wildcard,
+                    is_escape,
+                    lexer
+            );
+            bool already_added_var = false;
+            // Use the variable types to determine the possible_substr_types
+            for (int id : variable_types) {
+                auto& schema_type = lexer.m_id_symbol[id];
+                if (schema_type != "int" && schema_type != "float") {
+                    // LogSurgeon differentiates between all variable types. For example, LogSurgeon
+                    // might report thet types has#, userID, and int. However, CLP only supports
+                    // dict, int, and float variables. So there is no benefit in duplicating the
+                    // dict variable option for both has# and userID in the example.
+                    if (already_added_var) {
+                        continue;
+                    }
+                    already_added_var = true;
+                }
```

Comment:
Can we switch this code so that we not only consider the variable type, but also the encoding type here?

Let's assume we switch to the token-based approach mentioned in Grep.hpp, then this could be written as a method that returns one or more `VariableToken`s depending on whether a variable can be encoded or not. Then in this loop, we would loop over each of those `VariableToken`s can generate a `QueryLogtype` (`QueryInterpretation`). I think this is similar to what you were thinking of with the `TODO` in the unit tests.